BROOKLYN-264 Stop app while VM still being provisioned: vm is left

bostko opened the pull request:
- Initial Live Tests

andreaturli commented on software/base/src/test/java/org/apache/brooklyn/entity/EC2ExpungeLiveTest.java:
```diff
@@ -0,0 +1,128 @@
+        Entities.destroyCatching(app);
+        LOG.info("Time for expunging: {}", System.currentTimeMillis() - beginTime);
+
+        NodeMetadata nodeMetadata = Iterables.getFirst(((AWSEC2ComputeService) jcloudsLocation.getComputeService()).listNodesDetailsMatching(new Predicate<ComputeMetadata>() {
```
I think you can simplify with something like
```
import static com.google.common.base.Predicates.*;
import static org.jclouds.compute.predicates.NodePredicates.*;
import static com.google.common.collect.Iterables.*;

...

Iterable<? extends NodeMetadata> billedNodes = filter(jcloudsLocation.getComputeService().listNodesDetailsMatching(all()), and(withGroup(group), not(TERMINATED))));
```

bostko commented on software/base/src/test/java/org/apache/brooklyn/entity/EC2ExpungeLiveTest.java:
```diff
@@ -0,0 +1,128 @@
+     * </ul>
+     */
+    @Test(groups = {"Live"})
+    public void testExpunge() {
```
BTW the whole test passes for 3:45 seconds.
If we exclude the `listNodes` call, users will wait approximately 3 and a half minutes for expunge if they immediately click deploy and expunge.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java:
```diff
@@ -708,7 +710,21 @@ protected void doStop(ConfigBag parameters, Callable<StopMachineDetails<Integer>
+        // it may be safe to skip this wait and proceed with stop.
+        // Hesitant about parameterising this.
         Maybe<MachineLocation> machine = Machines.findUniqueMachineLocation(entity().getLocations());
+        if (machine.isAbsent()) {
```
This feels dangerous, for if start were to fail and not get a machine - we'd wait a long time, presumably.

aledsage commented on software/base/src/test/java/org/apache/brooklyn/entity/EC2ExpungeLiveTest.java:
```diff
@@ -0,0 +1,128 @@
+        try {
+            // Number of seconds we wait for an AWS EC2 instance to be created.
+            // The test succeeds if stop effector is called 200 seconds or more later
+            Thread.sleep(16 * 1000);
```
This kind of timing feels brittle to include in a test. But I guess we do need to wait some length of time, so it's not terminating it while it was still constructing the template (rather than in the middle of actually provisioning).

aledsage commented on software/base/src/test/java/org/apache/brooklyn/entity/EC2ExpungeLiveTest.java:
```diff
@@ -0,0 +1,128 @@
+            // The test succeeds if stop effector is called 200 seconds or more later
+            Thread.sleep(16 * 1000);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```
Don't bother catching and rethrowing. Just declare all your test methods as `throws Exception {`.

aledsage commented on software/base/src/test/java/org/apache/brooklyn/entity/EC2ExpungeLiveTest.java:
```diff
@@ -0,0 +1,128 @@
+
+        long beginTime = System.currentTimeMillis();
+        Entities.destroyCatching(app);
+        LOG.info("Time for expunging: {}", System.currentTimeMillis() - beginTime);
```
Use guava's `Stopwatch`, rather than this. Can also use `Duration.of(...)` to get a much nicer string representation of the duration.

aledsage commented:
My comments so far are bike-shedding.

Looking at the big picture, I need to think more about:

* How should we trigger the delete, so that it safely handles success and failure of provisioning? Options include:
  * As you've done here, have `stop()` wait for the machine to be set; but abort if start() fails.
  * Mark the entity's internal state as "requiring machine-release", and have the start() code immediately release the machine. But we'd still need `stop()` to wait for that to be done, otherwise we might go on to unmanage the entity and thus cancel the tasks.

* For speed:
  * Can we use the jclouds event bus to be told of the VM's id long before jclouds has returned from the provisioning call? @andreaturli what do you think?
  * Can we tell the `JcloudsLocation` instance that the provisioning has been aborted (this would allow us to avoid waiting-for-sshable, etc). Not quite sure how best to do that.

We can ignore the speed issue for now, unless there is a quick win that greatly speeds things up. The main thing is to get it functionally correct, and with clean code.

bostko commented:
Unfortunately this change dramatically slow down unit tests :(. 
Stopping EmptySoftwareProcess is a common task for most of the tests. Adding a five minute timeout on stop makes `mvn clean install` run for hours.

bostko commented:
I updated the PR however I do not know why tests in `rest-resources` are failing.

bostko commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java:
```diff
@@ -709,7 +713,24 @@ protected void doStop(ConfigBag parameters, Callable<StopMachineDetails<Integer>
+        /*Do not retry if non-jclouds location has started.*/
+        if (Boolean.TRUE.equals(entity().sensors().get(Sensors.newBooleanSensor("jclouds.provisioningStarted")))
+                && machine.isAbsent()) {
+            org.jclouds.util.Predicates2.retry(new Predicate<MachineLifecycleEffectorTasks>() {
```
What do you think about using `Predicates2.retry` across Apache Brooklyn. Can you suggest a guava alternative or what do you think about adding such function in `brooklyn-core`?

aledsage replied:
It feels like you should use our existing `org.apache.brooklyn.util.repeat.Repeater`. My personal opinion is that the builder pattern for the Repeater is easier to read, compared to the `Predicates2.retry` approach. The latter is probably only a good idea when you want to pass it to something that expects a `Predicate`.

bostko commented on software/base/src/test/java/org/apache/brooklyn/entity/EC2ExpungeLiveTest.java:
```diff
@@ -0,0 +1,144 @@
+     it will not retry and wait for a MachineLocation to appear.
+     */
+    @Test
+    public void testExpungeNonJcloudsLocation() throws Exception {
```
This test is for simple entities used in unit tests (have to improve naming here).
I am trying to add a similar Integration test for a BYON location.

bostko commented:
I think I have much better assertions now.

bostko commented on software/base/src/test/java/org/apache/brooklyn/entity/ExpungingJcloudsLocationLiveTest.java:
```diff
@@ -0,0 +1,187 @@
+     * </ul>
+     */
+    @Test(groups = {"Live"})
+    public void verifyJclousMachineIsExpungedWhenStoppedImmediatelyAfterStart() {
```
I have to add a test for one more case. Will see how straight it is to handle it.
When an exception occurs during deployment, in other words the execution goes here https://github.com/apache/brooklyn-server/blob/master/locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java#L1062-L1105
and `destroyOnFailure` is false then the machine will not be expunged.

bostko replied:
Supporting this case will need won't be one line change.
What  do you think about leaving this for a later PR?

aledsage commented on locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java:
```diff
@@ -735,7 +737,11 @@ protected MachineLocation obtainOnce(ConfigBag setup) throws NoMachinesAvailable
-                
+                Object callerContext = setup.get(JcloudsLocationConfig.CALLER_CONTEXT);
+                if (callerContext != null && callerContext instanceof Entity) {
+                    Entity callerContextEntity = (Entity)setup.get(JcloudsLocationConfig.CALLER_CONTEXT);
```
Minor: don't call `setup.get()` again - just use your `callerContext` variable.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java:
```diff
@@ -709,7 +712,26 @@ protected void doStop(ConfigBag parameters, Callable<StopMachineDetails<Integer>
+                            return Machines.findUniqueMachineLocation(entity().getLocations()).isPresent();
+                        }
+                    })
+                    .every(Duration.THIRTY_SECONDS)
```
Feels like strange times to use. Why only check every thirty seconds? That means we might take 29 seconds longer than we should have. The call to `getLocations()` should be cheap, so doing it every second seems fine.

aledsage replied:
And to be able to test it fast (in unit tests, rather than live tests), we'd want to set this very low. But let's not worry about that quite yet.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java:
```diff
@@ -709,7 +712,26 @@ protected void doStop(ConfigBag parameters, Callable<StopMachineDetails<Integer>
+                        }
+                    })
+                    .every(Duration.THIRTY_SECONDS)
+                    .limitTimeTo(Duration.FIVE_MINUTES)
```
Why wait for a hard-coded 5 minutes? Should be configurable. 5 minutes is not long enough on some clouds.

aledsage commented on software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java:
```diff
@@ -709,7 +712,26 @@ protected void doStop(ConfigBag parameters, Callable<StopMachineDetails<Integer>
         Maybe<MachineLocation> machine = Machines.findUniqueMachineLocation(entity().getLocations());
+        if (Boolean.TRUE.equals(entity().sensors().get(Attributes.JCLOUDS_PROVISIONING_STARTED))
+                && machine.isAbsent()) {
+            entity().sensors().set(Attributes.JCLOUDS_PROVISIONING_STARTED, null);
```
This name feels misleading. It's sort-of more about whether the provisioning is running, rather than started. I'd be inclined to use an enum to indicate the exact state. The value could be one of PROVISIONING, ABORTING, or null.

See later comment (not added yet!) about telling the jclouds provisioning to not do anything else (e.g. currently it will wait for ssh'able etc, which we can abort earlier).

bostko replied:
@aledsage do you have an idea how can I tell the stop task that the provisioning started other than the sensor flag. I feel like the code will need serious refactoring in order stop task to know earlier about the jclouds node.
The entity knows only about the machine location. It doesn't know about the LocationProvisioner.

bostko commented:
@aledsage do you have an idea how can I tell the stop task that the provisioning started other than the sensor flag. I feel like the code will need serious refactoring in order stop task to know earlier about the jclouds node.
The entity knows only about the machine location. It doesn't know about the LocationProvisioner.

Do you thing it is reasonable assigning the location provisioner to the sensor instead true or null?

bostko commented:
- I marked the sensor as "transient"
- I added a code which removes the sensor immidiately after a machine location is added
- I added an enum class for the `jclouds.provisioning.started` flag.<br>
  However I think with this code it could stay a boolean

aledsage commented:
I've been looking at how we can do this more generally (entirely in `MachineLifecycleEffectorTasks`, so it's not jclouds-specific). That makes the code easier to follow and easier to test. The former because the places we set the sensors is all in the same place; the latter because we can do it all with unit tests, rather than needing to provision VMs in live tests.

I have a commit that can go on top of this commit. I'll merge yours now, and then submit mine as a new PR.
